services/web-server/src/utils/requireContext.js

```diff
@@ -0,0 +1,51 @@
+const fs = require('fs');
+const path = require('path');
+const getCallerFile = require('get-caller-file');
+
+require.extensions['.graphql'] = function (module, filename) {
+  module.exports = fs.readFileSync(filename, 'utf8');
+};
+
+// Mock webpack's implementation of require.context
```

Comment:
Interesting. I was under the impression that this functionality is redundant on the backend, so there's not need to implement it?

Reply:
Without `require.context`, we still need to write something to loop over files in a directory and have them `require`d. `requireContext` does that. I went ahead and made the API look just like webpack's. Another way of writing this utility is to rewrite this function to return the full path of all files under a directory and have the file using this utility responsible of requiring the file. I think both implementations are fine except this one has the added bonus of being able to require the file directly.

Reply:
Is it possible to create an `index.js` file for the dir, import all the files from the dir there and then import that file instead of writing loops? (at least that is how I've it's usually done)

Reply:
> Is it possible to create an index.js file for the dir, import all the files from the dir there and then import that file

The 3 areas where we use `requireContext` are in `src/loader/index.js`, `src/resolvers/index.js` and `src/graphql/index.js`.

> instead of writing loops?

I'm not sure how to require the files without a loop. What did you have in mind?

Reply:
> I'm not sure how to require the files without a loop. Did you have something in mind?

*some file somewhere*

```
const loaders = require("./loaders");
```

*loaders/index.js*

```
exports.awesomeLoader = require("./loaders/awesomeLoader.js");
exports.fabulousLoader = require("./loaders/fabulousLoader.js");
exports.amazingLoader = require etc.
```

Reply:
https://github.com/taskcluster/taskcluster/blob/ef46fee21fa8d56ac990378534b53bee56f0ab9b/libraries/api/src/middleware/index.js#L5-L15
This is an example in our codebase.  I don't mind the re-implementation (with tests!) of `importer` here, although this is a pretty simple use-case for that function. It would be nice to have one way we do this throughout the codebase.

Reply:
Maybe this is OK as-is and fodder for a later clean-up?

Reply:
I wasn't arguing with Hassan here, I was trying to understand our use case. I see the `requireContext` to be rather hard to read and understand (and maybe it's just me) comparing to a series of `require`. I just wanted to understand the reason for using something more complex.

Reply:
I can see how questions like "why are we doing this instead of that" can come off as criticizing. It's just I am more focused on gaining expertise as a developer and learning from the team rather than criticizing you people :) I am also imagining I should become more productive if I understand our code base better

The current code is absolutely fine with me as long as it works, and especially if there's no explanation for me, we can just end this conversation right here :)

Reply:
Having to manually edit 3 `index.js` whenever we add/remove/rename a file in the `loader` directory for example is a cumbersome. We would usually need to modify 3 `index.js` files because adding a new file for example `loaders/test.js` also entails adding `resolvers/Test.js` and `graphql/Test.graphql`. Minimizing the mental overload of having to glue things together is what I was going for. I hope that clarifies the intent. Also, if you feel the `requireContext` is a bad name, I'm more than happy to change it to something else. I just took the name webpack uses since it's the same API.

Reply:
> I hope that clarifies the intent
It does, thank you!

> Also, if you feel the requireContext is a bad name, I'm more than happy to change it to something else
I seldom have problems with names :) In fact, it's something I don't usually even look at when doing a PR review